atlasdb-dbkvs-tests/src/test/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/oracle/OracleAlterTableIntegrationTest.java

```diff
@@ -101,6 +105,26 @@ public void after() {
         connectionSupplier.close();
     }
 
+    @Test
+    public void whenConfiguredAlterMetadataToMatchTableSchemaAndOldDataIsStillReadable() {
+        defaultKvs.createTable(TABLE_REFERENCE, EXPECTED_TABLE_METADATA.persistToBytes());
+        writeData(defaultKvs, ROW_1, TIMESTAMP_1);
+        assertThatDataCanBeRead(defaultKvs, DEFAULT_CELL_1, TIMESTAMP_1, DEFAULT_VALUE_1);
+
+        modifyMetadataToNotHaveOverflow();
+        // metadata is cached, thus we must create a new kvs to forcefully refresh the cache
+        KeyValueService badKvs = createKvs(DbKvsOracleTestSuite.getKvsConfig());
+        badKvs.putMetadataForTable(TABLE_REFERENCE, OLD_TABLE_METADATA.persistToBytes());
+        assertThatThrownBy(() -> fetchData(badKvs, DEFAULT_CELL_1, TIMESTAMP_1));
+
+        KeyValueService workingKvs = createKvs(CONFIG_WITH_ALTER);
+        workingKvs.createTable(TABLE_REFERENCE, EXPECTED_TABLE_METADATA.persistToBytes());
+        assertThatDataCanBeRead(defaultKvs, DEFAULT_CELL_1, TIMESTAMP_1, DEFAULT_VALUE_1);
+        assertThatMetadataHasOverflow();
+        assertThatCode(() -> writeData(defaultKvs, ROW_2, TIMESTAMP_2)).doesNotThrowAnyException();
+        assertThatDataCanBeRead(defaultKvs, DEFAULT_CELL_2, TIMESTAMP_2, DEFAULT_VALUE_2);
```

Comment:
non-actionable: Probably copied off the first test, and this isn't really captured by the name of the test (it is new data!), but it otherwise becomes too long I think.